Documentation: Document string formatting format

AtkinsSJ opened the pull request:
I kept missing this, so I wrote it. :^)

While doing so, I've learned that our implementation is pretty closely based on the C++ `std::formatter` specifiers, and that a lot of formatting is covered by TestFormat too. (Though, not hex float formatting, or plain chars.)

Maybe this would be better as a man page?

alimpfard commented on Documentation/StringFormatting.md:
```diff
@@ -0,0 +1,128 @@
+
+### Zero
+
+`0` pads the number with leading zeros.  
```
Extra trailing spaces? unbelievable!

alimpfard commented on Documentation/StringFormatting.md:
```diff
@@ -0,0 +1,128 @@
+
+### Type
+
+| Type     | Effect                | Example output           |
+|----------|-----------------------|--------------------------|
+| b        | binary                | `110`, `0b000110`        |
+| B        | binary uppercase      | `110`, `0B000110`        |
+| d        | decimal               | `42`, `+0000042`         |
+| o        | octal                 | `043`                    |
+| x        | hexadecimal           | `ff0`, `0x00000ff0`      |
+| X        | hexadecimal uppercase | `FF0`, `0X00000FF0`      |
+| c        | character             | `a`                      |
+| s        | string (default)      | `well, hello friends!`   |
+| p        | pointer               | `0xdeadc0de`             |
+| f        | float                 | `1.234`, `-inf`          |
+| a        | hex float             |                          |
+| A        | hex float uppercase   |                          |
+| hex-dump | hexadecimal dump      | `fdfdfdfd`, `3030    00` |
```
Just curious, does this render as a table in Serenity's markdown implementation?
Not a requirement for it to be compatible of course, just wondering :P

AtkinsSJ replied:
Not sure but I did look at LibMarkdown and it has table parsing code, I should test it properly though.

AtkinsSJ replied:
The answer is, no, it doesn't. 
![image](https://user-images.githubusercontent.com/222642/148132050-caa73f55-c40e-4c9b-86f2-a1bbadf0cea2.png)

There's clearly something happening there but it's not producing a proper table. :yakplus:

alimpfard commented on Documentation/StringFormatting.md:
````diff
@@ -0,0 +1,128 @@
+
+You can refer to the arguments by index, if you want to repeat one or change the order:
+```c++
+String::formatted("{3}{0}{1}", "a", "b", "c", "d") == "dab";
````
This actually won't compile, as the `"c"` argument is not used in the format string.

alimpfard replied:
(though the runtime is perfectly okay with evaluating this)

AtkinsSJ replied:
Thanks. Seems weird to me that `TestFormat.cpp` has test cases making sure things work that are bypassing the compile-time check so can't actually be used. If they're valid, the check should allow them; if they're not, why are we running tests to make sure they work?

alimpfard commented on Documentation/StringFormatting.md:
```diff
@@ -0,0 +1,128 @@
+
+`bool` values can be treated as an integer type (`0` or `1`), or as a string (`true` or `false`).
+
+`char` and `wchar` types can be treated as an integer type (their numeric value) or as a string.
```
I don't think we treat wide character strings as actual strings, and I doubt the `s` specifier knows how to print wide strings - that will likely result in a VERIFY (or a pointer being printed)

We should probably fix that ^^'

AtkinsSJ replied:
I wasn't meaning wide char _strings_ but just individual wchar_t types, though obviously I need to make that clearer! I might drop these two sentences since they're not very helpful.

trflynn89 commented on Documentation/StringFormatting.md:
````diff
@@ -0,0 +1,128 @@
+You can provide a custom `AK::Formatter<Foo>` class to format `Foo` values. For the simplest case where you already have a function that produces a string from your type, that would look like this:
+```c++
+
+namespace AK {
+template<>
+struct Formatter<Web::CSS::Selector> : Formatter<StringView> {
+    ErrorOr<void> format(FormatBuilder& builder, Web::CSS::Selector const& selector)
+    {
+        return Formatter<StringView>::format(builder, selector.serialize());
+    }
+};
+}
````
Quite a nitpick, but personally I'd rather see the struct declaration fully qualified, rather than add an enclosing namespace. I don't have a strong argument for it other than it vaguely sort of feels better than opening the AK namespace :grinning: 

```suggestion
template<>
struct AK::Formatter<Web::CSS::Selector> : Formatter<StringView> {
    ErrorOr<void> format(FormatBuilder& builder, Web::CSS::Selector const& selector)
    {
        return Formatter<StringView>::format(builder, selector.serialize());
    }
};
```

AtkinsSJ replied:
Oooh, I didn't know you could do that! Opening the namespace is what I've seen used and what I've copied from in the past, but this is a lot nicer, thanks!

trflynn89 commented on Documentation/StringFormatting.md:
```diff
@@ -0,0 +1,128 @@
+| x        | hexadecimal           | `ff0`, `0x00000ff0`      |
+| X        | hexadecimal uppercase | `FF0`, `0X00000FF0`      |
+| c        | character             | `a`                      |
+| s        | string (default)      | `well, hello friends!`   |
```
Dunno if it makes sense to say string is the default type. The default depends on the type of the value being formatted, no? e.g. integral values default to decimal.

AtkinsSJ replied:
I was making an assumption there, I see now that there's a distinct `Default` formatter mode apart from String, so I've updated the table accordingly.

AtkinsSJ commented:
Changes since last push:
- Adjusted invalid examples.
- Added a separate "default" row to the type specifiers table.
- Removed the two sentences about bools and chars which were more confusing than helpful. (Maybe a full breakdown of type specifiers per input type would be nice, but I don't feel like spending hours on that right now. ;^) )
- Used an `AK::` qualifier for the custom `Formatter` example instead of a namespace.
- Auto-formatted the file so it's nicer.
